atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java

```diff
@@ -0,0 +1,214 @@
+            return state.get();
+        }
+
+        while (true) {
+            try {
+                byte[] value = proposer.propose(seq, oldState == null ? null : PtBytes.toBytes(oldState));
+                // propose must never be null.  We only pass in null for things we know are agreed upon already.
+                Preconditions.checkNotNull(value, "Proposed value can't be null, but was in sequence %s", seq);
+                return ImmutableSequenceAndBound.of(seq, PtBytes.toLong(value));
+            } catch (PaxosRoundFailureException e) {
+                log.info("failed during propose", e);
+                long backoffTime = getRandomBackoffTime();
+                try {
+                    Thread.sleep(backoffTime);
+                } catch (InterruptedException e1) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        }
+    }
+
+    private Optional<SequenceAndBound> getLearnedState(long seq) {
+        if (seq <= PaxosAcceptor.NO_LOG_ENTRY) {
+            return Optional.of(ImmutableSequenceAndBound.of(PaxosAcceptor.NO_LOG_ENTRY, 0L));
+        }
+        List<PaxosLong> responses = PaxosQuorumChecker.collectQuorumResponses(
+                ImmutableList.copyOf(learners),
+                l -> getLearnedValue(seq, l),
+                1,
+                executor,
+                PaxosQuorumChecker.DEFAULT_REMOTE_REQUESTS_TIMEOUT_IN_SECONDS,
+                true);
+        if (responses.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(ImmutableSequenceAndBound.of(seq, responses.iterator().next().getValue()));
+    }
+
+    private static PaxosLong getLearnedValue(long seq, PaxosLearner learner) {
+        PaxosValue value = learner.getLearnedValue(seq);
+        if (value == null) {
+            throw new NoSuchElementException();
+        }
+        return ImmutablePaxosLong.of(PtBytes.toLong(value.getData()));
+    }
+
+    @Override
+    public synchronized void storeUpperLimit(long limit) throws MultipleRunningTimestampServiceError {
+        Preconditions.checkArgument(agreedState == null || limit >= agreedState.getBound());
+        long newSeq = agreedState == null ? 0 : agreedState.getSeqId() + 1;
+        while (true) {
+            try {
+                proposer.propose(newSeq, PtBytes.toBytes(limit));
+                PaxosValue value = knowledge.getLearnedValue(newSeq);
+                if (!value.getLeaderUUID().equals(proposer.getUUID())) {
+                    throw new MultipleRunningTimestampServiceError("concurrent proposal: " + value);
+                }
+                long newLimit = PtBytes.toLong(value.getData());
+                agreedState = ImmutableSequenceAndBound.of(newSeq, newLimit);
+                if (newLimit >= limit) {
+                    return;
+                }
+            } catch (PaxosRoundFailureException e) {
+                log.info("failed during propose", e);
```

Comment:
How does `Paxos proposal couldn't complete, because we could not connect to a quorum of nodes. We will retry in {}ms` sound? Agree that without familiarity with the code/paxos it looks concerning.